include/NAS2D/MathUtils.h

```diff
@@ -29,5 +30,30 @@ namespace NAS2D
 
 	int divideUp(int to_divide, int divisor);
 
+	
+
+/**
+ * \fn T rangeMap(T a, T b)
+ *
+ * Applies a linear transform to map an input range to an output range for built-in types.
+ *
+ * \param	valueToMap: The value to be mapped from the input range to the output range. Must start in the closed range [minInputRange,maxInputRange]
+ * \param	minInputRange: Smallest value to map from.
+ * \param	maxInputRange: Largest value to map from.
+ * \param	minOutputRange: Smallest value to map to.
+ * \param	maxOutputRange: Largest value to map to.
+ *
+ * \return	Returns the mapped value in the closed range [minOutputRange,maxOutputRange]
+ */
+template<typename T>
+T rangeMap(const T& valueToMap, const T& minInputRange, const T& maxInputRange, const T& minOutputRange, const T& maxOutputRange)
```

Comment:
For simple primitive types, which this is likely to be used with, we should drop the `const &`.

Reply:
I'm a little hesitant to drop the `const &` because it denotes at a glance that `valueToMap` will not be changed. As long as the type supports all the basic arithmetic operations it can be used here.

Reply:
The exact same thing can be said about passing by value. Plus that would remove the inefficiency of extra indirection.

Reply:
A passed-by-value parameter can be accidentally changed. A const value must be deliberately copied then changed. That is, it is indicative of a poor design or a malicious/dumb programmer is messing with something they probably shouldn't.

Reply:
*...accidentally changed inside the function...

Reply:
Run the code through a compiler and look at the assembly output. See how changing between `const &` and by-value affects the generated code. I recommend:
https://godbolt.org/